python/dgl/data/chem/alchemy.py

```diff
@@ -0,0 +1,303 @@
+                              Chem.rdchem.BondType.TRIPLE,
+                              Chem.rdchem.BondType.AROMATIC, None)
+                ])
+                bond_feats_dict['distance'].append(
+                    np.linalg.norm(geom[u] - geom[v]))
+
+        bond_feats_dict['e_feat'] = torch.FloatTensor(
+            bond_feats_dict['e_feat'])
+        bond_feats_dict['distance'] = torch.FloatTensor(
+            bond_feats_dict['distance']).reshape(-1, 1)
+
+        return bond_feats_dict
+
+    def mol_to_dgl(self, mol, self_loop=False):
+        """
+        Read sdf file and convert to dgl_graph
+        Args:
+            mol: molecule obj read from sdf
+            self_loop: Whetaher to add self loop
+        Returns:
+            g: DGLGraph
+            l: related labels
+        """
+
+        g = dgl.DGLGraph()
+
+        # add nodes
+        num_atoms = mol.GetNumAtoms()
+        atom_feats = self.alchemy_nodes(mol)
+        g.add_nodes(num=num_atoms, data=atom_feats)
+
+        # add edges
+        # The model we were interested assumes a complete graph.
+        # If this is not the case, do the code below instead
+        #
+        # for bond in mol.GetBonds():
+        #     u = bond.GetBeginAtomIdx()
+        #     v = bond.GetEndAtomIdx()
+        if self_loop:
+            g.add_edges(
+                [i for i in range(num_atoms) for j in range(num_atoms)],
+                [j for i in range(num_atoms) for j in range(num_atoms)])
+        else:
+            g.add_edges(
+                [i for i in range(num_atoms) for j in range(num_atoms - 1)], [
+                    j for i in range(num_atoms)
+                    for j in range(num_atoms) if i != j
+                ])
+
+        bond_feats = self.alchemy_edges(mol, self_loop)
+        g.edata.update(bond_feats)
+
+        return g
+
+    def __init__(self, mode='dev', transform=None, from_raw=False):
+        assert mode in ['dev', 'valid',
+                        'test'], "mode should be dev/valid/test"
+        self.mode = mode
+        self.transform = transform
+
+        # Construct the dgl graph from raw data or Use the preprossed data directly
+        self.from_raw = from_raw
+        file_dir = osp.join(get_download_dir(), './Alchemy_data')
+
+        if not from_raw:
+            file_name = "%s_processed" % (mode)
+        else:
+            file_name = "%s_single_sdf" % (mode)
+        self.file_dir = pathlib.Path(file_dir, file_name)
+
+        self.zip_file_path = pathlib.Path(file_dir, file_name + '.zip')
+        download(_urls['Alchemy'] + file_name + '.zip',
+                 path=str(self.zip_file_path))
+        if not os.path.exists(str(self.file_dir)):
+            archive = zipfile.ZipFile(self.zip_file_path)
+            archive.extractall(file_dir)
+            archive.close()
+
+        self._load()
+
+    def _load(self):
+        if self.mode == 'dev':
+            if not self.from_raw:
+                self.graphs = pickle.load(
```

Comment:
If would be better to use `with open(...) as f` than pure `open` so that files are guaranteed to be closed.